File: src/server/packets/PacketC2SSourceProperty.js
const PacketC2S = require("./PacketC2S");
const SourceProperty = require("../../rnet/SourceProperty");

/**
 * Client -> Server
 * ID = 0x34
 * Source Property
 * Change source properties
 */
class PacketC2SSourceProperty extends PacketC2S {
    parseData() {
        this._sourceID = this._buffer.readUInt8();
        this._propertyID = this._buffer.readUInt8();
        switch (this._propertyID) {
        case SourceProperty.PROPERTY_AUTO_OFF:
        case SourceProperty.PROPERTY_OVERRIDE_NAME:
            this._propertyValue = this._buffer.readUInt8() == 0x01;
            break;
        case SourceProperty.PROPERTY_AUTO_ON_ZONES:
            this._propertyValue = [];
            while (this._buffer.remaining() > 0) {
                this._propertyValue.push([this._buffer.readUInt8(), this._buffer.readUInt8()]);
            }
            break;
        default:
            this._propertyValue = false;
        }
    }

    getSourceID() {
        return this._sourceID;
    }

    getPropertyID() {
        return this._propertyID;
    }

    getPropertyValue() {
        return this._propertyValue;
    }

    getID() {
        return PacketC2SSourceProperty.ID;
    }
}

PacketC2SSourceProperty.ID = 0x34;

module.exports = PacketC2SSourceProperty;

File: src/RNetPi.Core/Packets/PacketC2SSourceProperty.cs
using RNetPi.Core.Constants;

namespace RNetPi.Core.Packets;

/// <summary>
/// Client -> Server
/// ID = 0x34
/// Source Property
/// Change source properties
/// Data:
///     (Unsigned Char) Source ID
///     (Unsigned Char) Property ID
///     (Variable) Property Value
/// </summary>
public class PacketC2SSourceProperty : PacketC2S
{
    public const byte ID = 0x34;

    public byte SourceID { get; private set; }
    public byte PropertyID { get; private set; }
    public object? PropertyValue { get; private set; }

    public PacketC2SSourceProperty(byte[] data) : base(data)
    {
    }

    public override byte GetID() => ID;

    protected override void ParseData()
    {
        SourceID = Reader.ReadByte();
        PropertyID = Reader.ReadByte();
        
        switch (PropertyID)
        {
            case SourceProperties.AutoOff:
            case SourceProperties.OverrideName:
                PropertyValue = Reader.ReadByte() == 0x01;
                break;
            case SourceProperties.AutoOnZones:
                var zones = new List<(byte ControllerID, byte ZoneID)>();
                while (Reader.BaseStream.Position < Reader.BaseStream.Length)
                {
                    var controllerID = Reader.ReadByte();
                    var zoneID = Reader.ReadByte();
                    zones.Add((controllerID, zoneID));
                }
                PropertyValue = zones.ToArray();
                break;
            default:
                PropertyValue = false;
                break;
        }
    }
}